panic on `tilt docker-prune`

### Steps to Reproduce
In tilt.build, run `tilt docker-prune`
(also repros in servantes and oneup)

### Current Behavior
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x18 pc=0x102adc9ac]

goroutine 1 [running]:
github.com/docker/distribution/reference.SplitHostname({0x0, 0x0})
	/Users/matt/go/src/github.com/tilt-dev/tilt/vendor/github.com/docker/distribution/reference/reference.go:183 +0x4c
github.com/docker/distribution/reference.TrimNamed(...)
	/Users/matt/go/src/github.com/tilt-dev/tilt/vendor/github.com/docker/distribution/reference/reference.go:323
github.com/tilt-dev/tilt/internal/container.RefSelector.AsNamedOnly(...)
	/Users/matt/go/src/github.com/tilt-dev/tilt/internal/container/selector.go:127
github.com/tilt-dev/tilt/internal/container.RefSet.LocalRef({{{0x0, 0x0}, 0x0}, {{0x0, 0x0}, {0x0, 0x0}, {0x0, 0x0}}})
	/Users/matt/go/src/github.com/tilt-dev/tilt/internal/container/reference.go:87 +0x60
github.com/tilt-dev/tilt/pkg/model.LocalRefSelectorsForManifests({0x14000b45500, 0x6, 0x8})
	/Users/matt/go/src/github.com/tilt-dev/tilt/pkg/model/manifest.go:392 +0x230
github.com/tilt-dev/tilt/internal/cli.(*dockerPruneCmd).run(0x1400068ae50, {0x1051ad668, 0x140000606c0}, {0x106974e90, 0x0, 0x0})
	/Users/matt/go/src/github.com/tilt-dev/tilt/internal/cli/docker_prune.go:67 +0x294
github.com/tilt-dev/tilt/internal/cli.addCommand.func1(0x140004bb400, {0x106974e90, 0x0, 0x0})
	/Users/matt/go/src/github.com/tilt-dev/tilt/internal/cli/cli.go:142 +0xac
github.com/spf13/cobra.(*Command).execute(0x140004bb400, {0x106974e90, 0x0, 0x0})
	/Users/matt/go/src/github.com/tilt-dev/tilt/vendor/github.com/spf13/cobra/command.go:860 +0x640
github.com/spf13/cobra.(*Command).ExecuteC(0x1400080f900)
	/Users/matt/go/src/github.com/tilt-dev/tilt/vendor/github.com/spf13/cobra/command.go:974 +0x410
github.com/spf13/cobra.(*Command).Execute(...)
	/Users/matt/go/src/github.com/tilt-dev/tilt/vendor/github.com/spf13/cobra/command.go:902
github.com/tilt-dev/tilt/internal/cli.Execute()
	/Users/matt/go/src/github.com/tilt-dev/tilt/internal/cli/cli.go:87 +0xcfc
main.main()
	/Users/matt/go/src/github.com/tilt-dev/tilt/cmd/tilt/main.go:19 +0x114
```


### Context
* I added some debug logging and confirmed `iTarg.Refs` is empty [here](https://github.com/tilt-dev/tilt/blob/d79b05e6a6f667e0a34599878a36394590fc8932/pkg/model/manifest.go#L389)
* I haven't dug too much further, but I'm guessing this is due to #5272 - `tilt docker-prune` uses the tiltfile load result directly, afaict and there are no reconcilers running, so iTarg.Refs is never populated
#### `tilt doctor` Output
```
$ tilt doctor
Tilt: v0.23.4-dev, built 2022-01-11
System: darwin-arm64
---
Docker
- Host: [default]
- Server Version: 20.10.11
- API Version: 1.41
- Builder: 2
- Compose Version: v2.2.1
---
Kubernetes
- Env: docker-for-desktop
- Context: docker-desktop
- Cluster Name: docker-desktop
- Namespace: default
- Container Runtime: docker
- Version: v1.22.4
- Cluster Local Registry: none
---
Thanks for seeing the Tilt Doctor!
Please send the info above when filing bug reports. 💗

The info below helps us understand how you're using Tilt so we can improve,
but is not required to ask for help.
---
Analytics Settings
--> (These results reflect your personal opt in/out status and may be overridden by an `analytics_settings` call in your Tiltfile)
- User Mode: opt-in
- Machine: b8b381be9003750ae6fe9bec80567735
- Repo: Rh53WceNSl4o/GBV8Qf6bQ==
```

--- a/integration/docker_prune/Tiltfile
+++ b/integration/docker_prune/Tiltfile
@@ -1,0 +1,14 @@
+# -*- mode: Python -*-
+
+include('../Tiltfile')
+
+docker_build('gcr.io/tilt-dev/image-for-prune', '.', dockerfile_contents='FROM nginx\n')
+
+k8s_yaml('pod.yaml')
+
+# we want to set unrealistically aggressive prune settings (i.e. will prune
+# EVERYTHING for this Tiltfile) for the purposes of testing
+# however, since `tilt ci` will run prune, we only set the opts when explicitly
+# running `tilt docker-prune` so we can control it for assert purposes
+if config.tilt_subcommand == 'docker-prune':
+    docker_prune_settings(keep_recent=0, max_age_mins=-1, num_builds=9999)

--- a/integration/docker_prune/pod.yaml
+++ b/integration/docker_prune/pod.yaml
@@ -1,0 +1,9 @@
+apiVersion: v1
+kind: Pod
+metadata:
+  name: site
+spec:
+  containers:
+    - name: site
+      image: gcr.io/tilt-dev/image-for-prune
+      imagePullPolicy: IfNotPresent

--- a/integration/docker_prune_test.go
+++ b/integration/docker_prune_test.go
@@ -4,9 +4,10 @@
 package integration
 
 import (
-	"context"
+	"io"
+	"os"
+	"strings"
 	"testing"
-	"time"
 
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
@@ -17,14 +18,26 @@ func TestCLI_DockerPrune(t *testing.T) {
 	defer f.TearDown()
 	f.SetRestrictedCredentials()
 
+	t.Log("Running `tilt ci` to trigger a build")
 	f.TiltCI()
+	t.Log("Running `tilt down` to stop running Pod so it can be pruned")
+	var outBuf strings.Builder
+	require.NoErrorf(t, f.tilt.Down(f.ctx, &outBuf),
+		"Error running `tilt down`. Output:\n%s\n", &outBuf)
 
-	ctx, cancel := context.WithTimeout(f.ctx, time.Minute)
-	defer cancel()
-
-	c := f.tilt.cmd(ctx, []string{"docker-prune", "--debug"}, nil)
+	outBuf.Reset()
+	c := f.tilt.cmd(f.ctx, []string{"docker-prune", "--debug"}, io.MultiWriter(&outBuf, os.Stdout))
 	t.Logf("Running command: %s", c.String())
-	out, err := c.CombinedOutput()
-	require.NoErrorf(t, err, "Error while running command. Output:\n%s\n", string(out))
-	assert.Contains(t, string(out), "[Docker Prune] removed 2 caches")
+	err := c.Run()
+	require.NoError(t, err, "Error while running command")
+	// use assert.True (instead of assert.Contains) - the cmd output is already
+	// output for debugging purposes and assert.Contains produces duplicative
+	// output (but hard to read due to embedded newlines)
+	out := outBuf.String()
+	assert.True(t,
+		strings.Contains(out, "- untagged:"),
+		"Image was not untagged")
+	assert.True(t,
+		strings.Contains(out, "- deleted: sha256:"),
+		"Image was not deleted")
 }

--- a/integration/docker_prune_test.go
+++ b/integration/docker_prune_test.go
@@ -1,0 +1,30 @@
+//go:build integration
+// +build integration
+
+package integration
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestCLI_DockerPrune(t *testing.T) {
+	f := newK8sFixture(t, "docker_prune")
+	defer f.TearDown()
+	f.SetRestrictedCredentials()
+
+	f.TiltCI()
+
+	ctx, cancel := context.WithTimeout(f.ctx, time.Minute)
+	defer cancel()
+
+	c := f.tilt.cmd(ctx, []string{"docker-prune", "--debug"}, nil)
+	t.Logf("Running command: %s", c.String())
+	out, err := c.CombinedOutput()
+	require.NoErrorf(t, err, "Error while running command. Output:\n%s\n", string(out))
+	assert.Contains(t, string(out), "[Docker Prune] removed 2 caches")
+}

--- a/internal/cli/docker_prune.go
+++ b/internal/cli/docker_prune.go
@@ -2,6 +2,7 @@ package cli
 
 import (
 	"context"
+	"os"
 	"strings"
 	"time"
 
@@ -55,9 +56,15 @@ func (c *dockerPruneCmd) run(ctx context.Context, args []string) error {
 	a.Incr("cmd.dockerPrune", nil)
 	defer a.Flush(time.Second)
 
-	// // (Most relevant output from dockerpruner is at the `debug` level)
-	// l := logger.NewLogger(logger.DebugLvl, os.Stdout)
-	// ctx = logger.WithLogger(ctx, l)
+	if !logger.Get(ctx).Level().ShouldDisplay(logger.VerboseLvl) {
+		// Docker Pruner filters output when nothing is pruned if not in verbose
+		// logging mode, which is suitable for when it runs in the background
+		// during `tilt up`, but we always want to include that for the CLI cmd
+		// N.B. we only override if we're not already showing verbose so that
+		// 	`--debug` flag isn't impacted
+		l := logger.NewLogger(logger.VerboseLvl, os.Stdout)
+		ctx = logger.WithLogger(ctx, l)
+	}
 
 	deps, err := wireDockerPrune(ctx, a, "docker-prune")
 	if err != nil {

--- a/internal/cli/docker_prune.go
+++ b/internal/cli/docker_prune.go
@@ -2,15 +2,18 @@ package cli
 
 import (
 	"context"
-	"os"
+	"strings"
 	"time"
 
 	"github.com/spf13/cobra"
 
 	"github.com/tilt-dev/tilt/internal/analytics"
+	"github.com/tilt-dev/tilt/internal/container"
 	ctrltiltfile "github.com/tilt-dev/tilt/internal/controllers/apis/tiltfile"
+	apitiltfile "github.com/tilt-dev/tilt/internal/controllers/core/tiltfile"
 	"github.com/tilt-dev/tilt/internal/docker"
 	"github.com/tilt-dev/tilt/internal/engine/dockerprune"
+	"github.com/tilt-dev/tilt/internal/k8s"
 	"github.com/tilt-dev/tilt/internal/tiltfile"
 	"github.com/tilt-dev/tilt/pkg/logger"
 	"github.com/tilt-dev/tilt/pkg/model"
@@ -22,12 +25,14 @@ type dockerPruneCmd struct {
 
 type dpDeps struct {
 	dCli docker.Client
+	kCli k8s.Client
 	tfl  tiltfile.TiltfileLoader
 }
 
-func newDPDeps(dCli docker.Client, tfl tiltfile.TiltfileLoader) dpDeps {
+func newDPDeps(dCli docker.Client, kCli k8s.Client, tfl tiltfile.TiltfileLoader) dpDeps {
 	return dpDeps{
 		dCli: dCli,
+		kCli: kCli,
 		tfl:  tfl,
 	}
 }
@@ -50,9 +55,9 @@ func (c *dockerPruneCmd) run(ctx context.Context, args []string) error {
 	a.Incr("cmd.dockerPrune", nil)
 	defer a.Flush(time.Second)
 
-	// (Most relevant output from dockerpruner is at the `debug` level)
-	l := logger.NewLogger(logger.DebugLvl, os.Stdout)
-	ctx = logger.WithLogger(ctx, l)
+	// // (Most relevant output from dockerpruner is at the `debug` level)
+	// l := logger.NewLogger(logger.DebugLvl, os.Stdout)
+	// ctx = logger.WithLogger(ctx, l)
 
 	deps, err := wireDockerPrune(ctx, a, "docker-prune")
 	if err != nil {
@@ -64,7 +69,10 @@ func (c *dockerPruneCmd) run(ctx context.Context, args []string) error {
 		return tlr.Error
 	}
 
-	imgSelectors := model.LocalRefSelectorsForManifests(tlr.Manifests)
+	imgSelectors, err := resolveImageSelectors(ctx, deps.kCli, &tlr)
+	if err != nil {
+		return err
+	}
 
 	dp := dockerprune.NewDockerPruner(deps.dCli)
 
@@ -72,4 +80,27 @@ func (c *dockerPruneCmd) run(ctx context.Context, args []string) error {
 	dp.Prune(ctx, tlr.DockerPruneSettings.MaxAge, tlr.DockerPruneSettings.KeepRecent, imgSelectors)
 
 	return nil
+}
+
+func resolveImageSelectors(ctx context.Context, kCli k8s.Client, tlr *tiltfile.TiltfileLoadResult) ([]container.RefSelector, error) {
+	registry := apitiltfile.DecideRegistry(ctx, kCli, tlr)
+	for _, m := range tlr.Manifests {
+		if err := m.InferImagePropertiesFromCluster(registry); err != nil {
+			return nil, err
+		}
+	}
+
+	imgSelectors := model.LocalRefSelectorsForManifests(tlr.Manifests)
+	if len(imgSelectors) != 0 && logger.Get(ctx).Level().ShouldDisplay(logger.DebugLvl) {
+		var sb strings.Builder
+		for _, is := range imgSelectors {
+			sb.WriteString("  - ")
+			sb.WriteString(is.RefFamiliarString())
+			sb.WriteRune('\n')
+		}
+
+		logger.Get(ctx).Debugf("Running Docker Prune for images:\n%s", sb.String())
+	}
+
+	return imgSelectors, nil
 }

--- a/internal/cli/wire_gen.go
+++ b/internal/cli/wire_gen.go
@@ -166,7 +166,7 @@ func wireDockerPrune(ctx context.Context, analytics2 *analytics.TiltAnalytics, s
 	processExecer := localexec.NewProcessExecer(localexecEnv)
 	defaults := _wireDefaultsValue
 	tiltfileLoader := tiltfile.ProvideTiltfileLoader(analytics2, plugin, versionPlugin, configPlugin, dockerComposeClient, webHost, processExecer, defaults, env)
-	cliDpDeps := newDPDeps(switchCli, tiltfileLoader)
+	cliDpDeps := newDPDeps(switchCli, client, tiltfileLoader)
 	return cliDpDeps, nil
 }
 

--- a/internal/controllers/core/tiltfile/reconciler.go
+++ b/internal/controllers/core/tiltfile/reconciler.go
@@ -318,7 +318,7 @@ func (r *Reconciler) run(ctx context.Context, nn types.NamespacedName, tf *v1alp
 // apiserver.
 func (r *Reconciler) handleLoaded(ctx context.Context, nn types.NamespacedName, tf *v1alpha1.Tiltfile, entry *BuildEntry, tlr *tiltfile.TiltfileLoadResult) error {
 	// TODO(nick): Rewrite to handle multiple tiltfiles.
-	err := updateOwnedObjects(ctx, r.ctrlClient, nn, tf, tlr, r.engineMode, r.decideRegistry(ctx, tlr), r.defaultK8sConnection())
+	err := updateOwnedObjects(ctx, r.ctrlClient, nn, tf, tlr, r.engineMode, DecideRegistry(ctx, r.k8sClient, tlr), r.defaultK8sConnection())
 	if err != nil {
 		// If updating the API server fails, just return the error, so that the
 		// reconciler will retry.
@@ -404,16 +404,16 @@ func (r *Reconciler) enqueueTriggerQueue(obj client.Object) []reconcile.Request 
 	return requests
 }
 
-// decideRegistry returns the image registry we should use; if detected, a pre-configured
+// DecideRegistry returns the image registry we should use; if detected, a pre-configured
 // local registry; otherwise, the registry specified by the user via default_registry.
 // Otherwise, we'll return the zero value of `s.defaultReg`, which is an empty registry.
 // It has side-effects (a log line) and so should only be called once.
-func (r *Reconciler) decideRegistry(ctx context.Context, tlr *tiltfile.TiltfileLoadResult) container.Registry {
+func DecideRegistry(ctx context.Context, kCli k8s.Client, tlr *tiltfile.TiltfileLoadResult) container.Registry {
 	if tlr.Orchestrator() != model.OrchestratorK8s {
 		return tlr.DefaultRegistry
 	}
 
-	registry := r.k8sClient.LocalRegistry(ctx)
+	registry := kCli.LocalRegistry(ctx)
 
 	if !registry.Empty() {
 		// If we've found a local registry in the cluster at run-time, use that

--- a/internal/engine/dockerprune/docker_pruner.go
+++ b/internal/engine/dockerprune/docker_pruner.go
@@ -13,7 +13,6 @@ import (
 	"github.com/docker/docker/api/types/filters"
 
 	"github.com/tilt-dev/tilt/internal/container"
-
 	"github.com/tilt-dev/tilt/pkg/model"
 
 	"github.com/tilt-dev/tilt/internal/engine/buildcontrol"
@@ -302,7 +301,7 @@ func (dp *DockerPruner) sufficientVersionError() error {
 }
 
 func prettyPrintImagesPruneReport(report types.ImagesPruneReport, l logger.Logger) {
-	if len(report.ImagesDeleted) == 0 && !l.Level().ShouldDisplay(logger.DebugLvl) {
+	if len(report.ImagesDeleted) == 0 && !l.Level().ShouldDisplay(logger.InfoLvl) {
 		return
 	}
 
@@ -326,7 +325,7 @@ func prettyStringImgDeleteItem(img types.ImageDeleteResponseItem) string {
 }
 
 func prettyPrintCachePruneReport(report *types.BuildCachePruneReport, l logger.Logger) {
-	if len(report.CachesDeleted) == 0 && !l.Level().ShouldDisplay(logger.DebugLvl) {
+	if len(report.CachesDeleted) == 0 && !l.Level().ShouldDisplay(logger.InfoLvl) {
 		return
 	}
 
@@ -338,7 +337,7 @@ func prettyPrintCachePruneReport(report *types.BuildCachePruneReport, l logger.L
 }
 
 func prettyPrintContainersPruneReport(report types.ContainersPruneReport, l logger.Logger) {
-	if len(report.ContainersDeleted) == 0 && !l.Level().ShouldDisplay(logger.DebugLvl) {
+	if len(report.ContainersDeleted) == 0 && !l.Level().ShouldDisplay(logger.InfoLvl) {
 		return
 	}
 

--- a/internal/engine/dockerprune/docker_pruner.go
+++ b/internal/engine/dockerprune/docker_pruner.go
@@ -13,6 +13,7 @@ import (
 	"github.com/docker/docker/api/types/filters"
 
 	"github.com/tilt-dev/tilt/internal/container"
+
 	"github.com/tilt-dev/tilt/pkg/model"
 
 	"github.com/tilt-dev/tilt/internal/engine/buildcontrol"
@@ -301,7 +302,7 @@ func (dp *DockerPruner) sufficientVersionError() error {
 }
 
 func prettyPrintImagesPruneReport(report types.ImagesPruneReport, l logger.Logger) {
-	if len(report.ImagesDeleted) == 0 && !l.Level().ShouldDisplay(logger.InfoLvl) {
+	if len(report.ImagesDeleted) == 0 && !l.Level().ShouldDisplay(logger.VerboseLvl) {
 		return
 	}
 
@@ -325,7 +326,7 @@ func prettyStringImgDeleteItem(img types.ImageDeleteResponseItem) string {
 }
 
 func prettyPrintCachePruneReport(report *types.BuildCachePruneReport, l logger.Logger) {
-	if len(report.CachesDeleted) == 0 && !l.Level().ShouldDisplay(logger.InfoLvl) {
+	if len(report.CachesDeleted) == 0 && !l.Level().ShouldDisplay(logger.VerboseLvl) {
 		return
 	}
 
@@ -337,7 +338,7 @@ func prettyPrintCachePruneReport(report *types.BuildCachePruneReport, l logger.L
 }
 
 func prettyPrintContainersPruneReport(report types.ContainersPruneReport, l logger.Logger) {
-	if len(report.ContainersDeleted) == 0 && !l.Level().ShouldDisplay(logger.InfoLvl) {
+	if len(report.ContainersDeleted) == 0 && !l.Level().ShouldDisplay(logger.VerboseLvl) {
 		return
 	}
 

